benchmarks/benchmark/tools/locust-load-inference/locust-docker/locust-tasks/tasks.py

```diff
@@ -300,3 +306,78 @@ def total_content_length():
             Add a route to the Locust web app, where we can see the total content-length
             """
             return local_metric_collector.json_dump_report()
+
+class GrpcUser(User):
+    abstract = True
+    stub_class = None
+
+    def __init__(self, environment):
+        super().__init__(environment)
+        for attr_value, attr_name in ((self.host, "host"), (self.stub_class, "stub_class")):
+            if attr_value is None:
+                raise LocustError(f"You must specify the {attr_name}.")
+
+        self._channel = grpc.insecure_channel(self.host)
+        interceptor = LocustInterceptor(environment=environment)
+        self._channel = grpc.intercept_channel(self._channel, interceptor)
+
+        self.stub = self.stub_class(self._channel)
+
+class JetStreamUser(GrpcUser):
+    stub_class = jetstream_pb2_grpc.OrchestratorStub
+
+    @task
+    def jetstream_grpc_infer(self):
+        prompt = test_data[random.randrange(0, len(test_data))]
+        request = jetstream_pb2.DecodeRequest(
+            additional_text=prompt,
+            priority=0,
+            max_tokens=model_params["max_output_len"],
+        )
+        logging.info(f"Prompt: {prompt}")
+        #return values format is from the interceptor, which makes the actual call
+        output, ttft, response_time = self.stub.Decode(request)
+        logging.info(f"Response: {output}")
+
+        number_of_input_tokens = len(tokenizer.encode(prompt))
+        number_of_output_tokens = len(tokenizer.encode(output))
+        send_metrics(number_of_input_tokens, number_of_output_tokens, response_time,1)
+
+
+class LocustInterceptor(ClientInterceptor):
+    def __init__(self, environment, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self.env = environment
+
+    def intercept(
+        self,
+        method: Callable,
+        request_or_iterator: Any,
+        call_details: grpc.ClientCallDetails,
+    ):
+        response = None
+        exception = None
+        start_perf_counter = time.perf_counter()
+        response_length = 0
+        responses = method(request_or_iterator, call_details)
+        output = ""
+        response_length = 0
+        ttft = 0
+        for response in responses:
+            if ttft == 0:
+                ttft = time.perf_counter() - start_perf_counter
```

Comment:
I'm not sure I understand how ttft is being calculated here. Can you expand? Also, would be good to add a comment on what this does.

Reply:
For sure! Adding a comment, but also explaining here. We stream the response, so as each chunk is sent back to us, we create the full output. When we receive our first token back, we calculate that t-delta to calculate TTFT. 

The assumption is that each chunk returned is a single token. But Jetstream makes a similar assumption: https://github.com/google/JetStream/blob/main/benchmarks/benchmark_serving.py#L381


Actual output from benchmark logs:  `response_time 5931.7820890573785; ttft:170.58807401917875`

Reply:
Makes sense. Thanks for the clarification!